File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hm = {}
        l = 0
        maxLength = 0

        for r, char in enumerate(s):
            if char not in hm or hm[char] < l: 
                hm[char] = r
            else:
                l = hm[char] + 1 
                hm[char] = r

            maxLength = max(maxLength, r - l + 1)
        
        return maxLength


            

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-3.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;

        int maxProfit = 0;
        int currentMin = prices[0];
        
        for (int price : prices) {
            currentMin = Math.min(currentMin, price);
            maxProfit = Math.max(maxProfit, price - currentMin);
        }

        return maxProfit;
    }
}
